File: November2021/C++/Multiply_Strings.cpp
// author : shyam2520 
// https://leetcode.com/problems/multiply-strings/
class Solution {
public:
    string multiply(string num1, string num2) {
        int m=num1.size(),n=num2.size();
        vector<int> res(n+m,0);
        for(int i=m-1;i>=0;i--)
        {
            for(int j=n-1;j>=0;j--)
            {
                int mul=(int)(num2[j]-'0')*(int)(num1[i]-'0');
                int sum=mul+res[i+j+1];
                res[i+j+1]=sum%10;
                res[i+j]+=sum/10;
            }
        }
        string ans="";
        int i=0;
        while(i<n+m)
        {
            if(res[i]) 
                break;
            i++;
        }
        
        for(;i<n+m;i++)
            ans+=res[i]+'0';
       
        if(!ans.size())
            return "0";
        return ans;
    }
};

File: November2021/MultiplyStrings.java
class Solution {

	public String multiply(String num1, String num2) {

		int m = num1.length();
		int n = num2.length();

		// Base Conditions
		if (m == 0 || n == 0 || "0".equals(num1) || "0".equals(num2)) {
			return "0";
		}
		if ("1".equals(num1)) {
			return num2;
		}
		if ("1".equals(num2)) {
			return num1;
		}

		// Result can be maximum of length M + N.
		// For example 99 * 999 = 98901 (Result is of length 5)
		int[] result = new int[m + n];

		for (int i = m - 1; i >= 0; i--) {
			for (int j = n - 1; j >= 0; j--) {
				int product = (num1.charAt(i) - '0') * (num2.charAt(j) - '0');
				// Adding previous values in result array into this product.
				product += result[i + j + 1];

				// Adding the new product into the result array
				result[i + j + 1] = product % 10;
				result[i + j] += product / 10;
			}
		}

		StringBuilder sb = new StringBuilder();
		for (int r : result) {
			if (sb.length() == 0 && r == 0) {
				continue;
			}
			sb.append(r);
		}

		return sb.toString();
	}

}